genesis/engine/couplers/sap_coupler.py

```diff
@@ -0,0 +1,1834 @@
+    def check_initial_exact_linesearch_convergence(self):
+        fem_solver = self.fem_solver
+        for i_b in ti.ndrange(self._B):
+            if not self.batch_linesearch_active[i_b]:
+                continue
+            self.batch_linesearch_active[i_b] = self.linesearch_state[i_b].dell_dalpha > 0.0
+        # When tolerance is small but gradient norm is small, take step 1.0 and end
+        for i_b, i_v in ti.ndrange(self._B, fem_solver.n_vertices):
+            if not self.batch_linesearch_active[i_b]:
+                continue
+            if (
+                -self.linesearch_state[i_b].m
+                < self._sap_convergence_atol + self._sap_convergence_rtol * self.linesearch_state[i_b].prev_energy
+            ):
+                self.v[i_b, i_v] = self.linesearch_state_v[i_b, i_v].x_prev + self.pcg_state_v[i_b, i_v].x
+        for i_b, i_v in ti.ndrange(self._B, fem_solver.n_vertices):
+            if not self.batch_linesearch_active[i_b]:
+                continue
+            if (
+                -self.linesearch_state[i_b].m
+                < self._sap_convergence_atol + self._sap_convergence_rtol * self.linesearch_state[i_b].prev_energy
+            ):
+                self.batch_linesearch_active[i_b] = False
+                self.linesearch_state[i_b].step_size = 1.0
+
+    def one_linesearch_iter(self, f: ti.i32):
+        self.update_velocity_linesearch()
+        self.compute_total_energy(f, self.linesearch_state.energy)
+        self.check_linesearch_convergence()
+
+    @ti.kernel
+    def update_velocity_linesearch(self):
+        fem_solver = self.fem_solver
+
+        # update vel
+        for i_b, i_v in ti.ndrange(self._B, fem_solver.n_vertices):
+            if not self.batch_linesearch_active[i_b]:
+                continue
+            self.v[i_b, i_v] = (
+                self.linesearch_state_v[i_b, i_v].x_prev
+                + self.linesearch_state[i_b].step_size * self.pcg_state_v[i_b, i_v].x
+            )
+
+    @ti.kernel
+    def check_linesearch_convergence(self):
+        # check condition
+        for i_b in ti.ndrange(self._B):
+            if not self.batch_linesearch_active[i_b]:
+                continue
+            self.batch_linesearch_active[i_b] = (
+                self.linesearch_state[i_b].energy
+                > self.linesearch_state[i_b].prev_energy
+                + self._linesearch_c * self.linesearch_state[i_b].step_size * self.linesearch_state[i_b].m
+            )
+            if not self.batch_linesearch_active[i_b]:
+                continue
+            self.linesearch_state[i_b].step_size *= self._linesearch_tau
+
+    def linesearch(self, f: ti.i32):
+        """
+        Note
+        ------
+        https://en.wikipedia.org/wiki/Backtracking_line_search#Algorithm
+        """
+        self.init_linesearch(f)
+        for i in range(self._n_linesearch_iterations):
+            self.one_linesearch_iter(f)
+
+    def exact_linesearch(self, f: ti.i32):
```

Comment:
Could you add a short docstring that explain the difference compared to `linesearch`. Why both options are available?